File: test2.py
import numpy as np
import sobol_seq
import pyDOE  # Updated to pyDOE2, as pyDOE has been replaced by pyDOE2
from deap import base, creator, tools
from scipy.optimize import differential_evolution
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import ParameterSampler
from skopt import BayesSearchCV
import torch
import torch.nn as nn
from concurrent.futures import ThreadPoolExecutor
from sklearn.datasets import load_iris, load_digits
from sklearn.cluster import KMeans
import random

# Step 1: Define Neural Network Model
class NeuralNet(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim, activation_fn):
        super(NeuralNet, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, output_dim)
        self.activation_fn = activation_fn()

    def forward(self, x):
        x = self.activation_fn(self.fc1(x))
        x = self.fc2(x)
        return x

# Step 2: Define Sampling Methods
# Latin Hypercube Sampling (LHS)
def lhs_sampling(num_samples, dims):
    return pyDOE.lhs(dims, samples=num_samples)

# Sobol Sequences
def sobol_sampling(num_samples, dims):
    return sobol_seq.i4_sobol_generate(dims, num_samples)

# Random Forest Hyperparameter Sampling
def random_forest_sampling(param_grid, num_samples):
    samples = list(ParameterSampler(param_grid, n_iter=num_samples))
    return samples

# Genetic Algorithm Sampling
def genetic_algorithm_sampling(num_samples, dims):
    creator.create("FitnessMax", base.Fitness, weights=(1.0,))
    creator.create("Individual", list, fitness=creator.FitnessMax)

    toolbox = base.Toolbox()
    toolbox.register("attr_float", np.random.random)
    toolbox.register("individual", tools.initRepeat, creator.Individual, toolbox.attr_float, n=dims)
    toolbox.register("population", tools.initRepeat, list, toolbox.individual)

    population = toolbox.population(n=num_samples)
    population_np = np.array([ind for ind in population])
    return population_np

    return population

# Differential Evolution
def func_to_optimize(x):
    return np.sum(x ** 2)  # Example of a function, can be modified

def differential_evolution_sampling(bounds):
    result = differential_evolution(func_to_optimize, bounds)
    return result.x

# Bayesian Optimization Sampling
def bayesian_optimization_sampling(model, param_bounds, num_samples):
    optimizer = BayesSearchCV(estimator=model, search_spaces=param_bounds, n_iter=num_samples)
    optimizer.fit(np.random.rand(100, param_bounds.shape[0]), np.random.rand(100))  # Placeholder for random data
    return optimizer

# Step 3: Split Search Space into Convex Regions
def split_search_space(search_space, num_regions):
    kmeans = KMeans(n_clusters=num_regions)
    labels = kmeans.fit_predict(search_space)
    regions = [search_space[labels == i] for i in range(num_regions)]
    return regions

# Step 4: Define Objective Function for IPOPT (Example using MSE for Regression)
def objective_function(params, model, data, target):
    data = torch.Tensor(data)
    target = torch.Tensor(target)
    model_input = torch.Tensor(params).view(-1, data.size(1))
    output = model(model_input)
    loss = nn.MSELoss()(output, target)
    return loss.item()

# Step 5: IPOPT solver is omitted as pyipopt isn't widely supported
# We'll use a placeholder here

# Step 6: Parallel Processing with ThreadPoolExecutor
def parallel_processing(sampling_method, model, data, target, bounds, num_workers=4):
    samples = sampling_method()
    regions = split_search_space(samples, num_regions=5)

    with ThreadPoolExecutor(max_workers=num_workers) as executor:
        futures = [executor.submit(func_to_optimize, region) for region in regions]
        results = [f.result() for f in futures]
    
    return results

# Step 7: Define Problems with Relevant Data and Models
# XOR Problem (2 inputs, 2 hidden, 1 output)
XOR_data = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
XOR_target = np.array([0, 1, 1, 0])

# Iris Problem (4 inputs, 4 hidden, 3 output)
iris = load_iris()
X_iris, y_iris = iris.data, iris.target

# Heart Disease (32 inputs, 10 hidden, 1 output)
Heart_data = np.random.randn(100, 32)  # Placeholder for real heart dataset
Heart_target = np.random.randint(2, size=100)

# MNIST (784 inputs, 10 hidden, 10 output)
mnist = load_digits()
X_mnist, y_mnist = mnist.data, mnist.target

# Step 8: Define Search Space Bounds
XOR_bounds = [(0, 1)] * 2
iris_bounds = [(0, 7)] * 4
heart_bounds = [(0, 1)] * 32
mnist_bounds = [(0, 16)] * 784

# Step 9: Choose Activation Functions (Tanh, ReLU, ELU)
activation_functions = [nn.Tanh, nn.ReLU, nn.ELU]

# Step 10: Run Each Sampling Method and Neural Network Model
def run_sampling_method(sampling_method, model, data, target, bounds, num_workers=4):
    results = parallel_processing(sampling_method, model, data, target, bounds, num_workers=num_workers)
    return results

# Step 11: Neural Network Models for XOR, Iris, Heart, MNIST Problems
def run_experiments():
    xor_model = NeuralNet(input_dim=2, hidden_dim=2, output_dim=1, activation_fn=nn.Tanh)
    iris_model = NeuralNet(input_dim=4, hidden_dim=4, output_dim=3, activation_fn=nn.Tanh)
    heart_model = NeuralNet(input_dim=32, hidden_dim=10, output_dim=1, activation_fn=nn.ReLU)
    mnist_model = NeuralNet(input_dim=784, hidden_dim=10, output_dim=10, activation_fn=nn.ELU)

    # Run sampling for XOR problem
    xor_results = run_sampling_method(lambda: lhs_sampling(100, 2), xor_model, XOR_data, XOR_target, XOR_bounds)
    
    # Run sampling for Iris problem
    iris_results = run_sampling_method(lambda: sobol_sampling(100, 4), iris_model, X_iris, y_iris, iris_bounds)
    
    # Run sampling for Heart problem
    heart_results = run_sampling_method(lambda: genetic_algorithm_sampling(100, 32), heart_model, Heart_data, Heart_target, heart_bounds)
    
    # Run sampling for MNIST problem
    mnist_results = run_sampling_method(lambda: differential_evolution_sampling(mnist_bounds), mnist_model, X_mnist, y_mnist, mnist_bounds)

    return xor_results, iris_results, heart_results, mnist_results

# Step 12: Compare Results
def compare_results(results_dict):
    for method, results in results_dict.items():
        print(f"Results for {method}:")
        print(results)

# Step 13: Main Execution
if __name__ == "__main__":
    # Define the results for comparison
    xor_results, iris_results, heart_results, mnist_results = run_experiments()

    # Compare all results for each problem
    results_dict = {
        "XOR_LHS": xor_results,
        "Iris_Sobol": iris_results,
        "Heart_GA": heart_results,
        "MNIST_DE": mnist_results,
    }
    
    compare_results(results_dict)

File: test2.jl
using Pkg
Pkg.add("Flux")
Pkg.add("StatsBase")
Pkg.add("Clustering")
Pkg.add("Sobol")
Pkg.add("MLDatasets")
Pkg.add("Plots")
Pkg.add("JuMP")
Pkg.add("Ipopt")
Pkg.add("Random")
Pkg.add("Distributions")
Pkg.add("GaussianProcesses")
Pkg.add("BayesianOptimization")
Pkg.add("Hyperband")
Pkg.add("DifferentialEvolution")
Pkg.add("GeneticAlgorithms")
Pkg.add("ParticleSwarmOptimization")

using Flux, StatsBase, Clustering, Sobol, MLDatasets, Plots, JuMP, Ipopt, Random, Distributions, GaussianProcesses, BayesianOptimization, Hyperband, DifferentialEvolution, GeneticAlgorithms, ParticleSwarmOptimization

# Define the architecture for the neural networks
function build_network(input_dim, hidden_dim, output_dim, activation_fn)
    return Chain(Dense(input_dim, hidden_dim, activation_fn), Dense(hidden_dim, output_dim))
end

# Define sampling methods
function lhs_sampling(num_samples, dims)
    return lhssample(dims, num_samples)
end

function sobol_sampling(num_samples, dims)
    return sobol(num_samples, dims)
end

function random_forest_sampling(num_samples, dims)
    # Simulate random forest hyperparameter optimization
    # Here we'll just generate random samples for simplicity
    return rand(Uniform(0, 1), num_samples, dims)
end

function genetic_algorithm_sampling(num_samples, dims)
    # Generate random population for Genetic Algorithm
    return rand(Uniform(0, 1), num_samples, dims)
end

function particle_swarm_sampling(num_samples, dims)
    # Simulate Particle Swarm Sampling by generating random samples
    return rand(Uniform(0, 1), num_samples, dims)
end

function bayesian_optimization(num_samples, dims)
    # Simulate Bayesian Optimization by generating random samples
    return rand(Uniform(0, 1), num_samples, dims)
end

function adaptive_sampling(num_samples, dims)
    # Simulate Adaptive Sampling (LHS followed by refinement)
    return rand(Uniform(0, 1), num_samples, dims)
end

# Step 2: Split Search Space into Convex Regions
function split_search_space(search_space, num_regions)
    labels = kmeans(Matrix(search_space), num_regions).assignments
    regions = [search_space[labels .== i, :] for i in 1:num_regions]
    return regions
end

# Step 3: Optimization using IPOPT
function solve_with_ipopt(bounds, objective_function)
    n_dims = length(bounds)
    model_opt = Model(Ipopt.Optimizer)

    @variable(model_opt, bounds[i][1] <= x[i=1:n_dims] <= bounds[i][2])
    @objective(model_opt, Min, objective_function(x))
    
    optimize!(model_opt)
    return value.(x)
end

# Define a simple loss function for optimization
function objective_function(x)
    return sum(x .^ 2)
end

# Step 4: Train the network using SGD
function train_with_sgd(model, data, target, epochs, lr)
    loss(x, y) = Flux.mse(model(x), y)
    opt = Flux.Descent(lr)
    for epoch in 1:epochs
        Flux.train!(loss, params(model), [(data, target)], opt)
    end
end

# Define datasets
function load_datasets()
    XOR_data = [0 0; 0 1; 1 0; 1 1]
    XOR_target = [0, 1, 1, 0]

    iris_data, iris_target = MLDatasets.Iris()

    heart_data = rand(100, 32)
    heart_target = rand(0:1, 100)

    mnist_data, mnist_target = MLDatasets.MNIST.traindata()

    return XOR_data, XOR_target, iris_data, iris_target, heart_data, heart_target, mnist_data, mnist_target
end

# Step 5: Run all sampling methods and save plots
function run_sampling_and_plot(methods, model, data, target, bounds, dataset_name, activation_fn_name)
    results = []
    for method in methods
        search_space = method()
        regions = split_search_space(search_space, num_regions=5)
        for region in regions
            optimal = solve_with_ipopt(bounds, objective_function)
            push!(results, optimal)
        end
    end

    # Train model using SGD for comparison
    train_with_sgd(model, data, target, epochs=100, lr=0.01)
    
    # Plot results
    plot(results, label="Sampling Methods")
    savefig("results/$(dataset_name)_$(activation_fn_name).png")
end

# Main Execution
function main()
    XOR_data, XOR_target, iris_data, iris_target, heart_data, heart_target, mnist_data, mnist_target = load_datasets()

    datasets = [
        ("XOR", XOR_data, XOR_target, [(0, 1), (0, 1)]),
        ("Iris", iris_data, iris_target, [(0, 7), (0, 7), (0, 7), (0, 7)]),
        ("Heart", heart_data, heart_target, [(0, 1) for _ in 1:32]),
        ("MNIST", mnist_data, mnist_target, [(0, 16) for _ in 1:784])
    ]

    methods = [
        lhs_sampling, sobol_sampling, random_forest_sampling, genetic_algorithm_sampling,
        particle_swarm_sampling, bayesian_optimization, adaptive_sampling
    ]

    activation_functions = [Flux.tanh, Flux.relu, Flux.elu]

    for (dataset_name, data, target, bounds) in datasets
        for activation_fn in activation_functions
            model = build_network(size(data, 2), 10, size(target, 1), activation_fn)
            run_sampling_and_plot(methods, model, data, target, bounds, dataset_name, string(activation_fn))
        end
    end
end

main()
